Fix blanking windows not being drawn into on macOS

prewettg opened the pull request:
Multiple windows on macOS currently flicker: as one is drawn into, the other one is blanked to the NSView color. This is caused by NSOpenGLContext -clearDrawable, which not only disassociates the view from the context as the documentation says, but also blanks the view before disassociating from it. This seems to be necessary when the view is resized, but is not desirable in other situations. This PR only calls -clearDrawable when it can detect that the NSView has resized, otherwise it calls -setView: if necessary and -update.

prewettg commented:
CI is failing on Windows, but it seems to be because the compiler runs out of heap space on gltf_viewer, which hasn't changed in this PR. The only change was in an Objective-C++ file, so there should be no impact on Windows at all.

prideout commented on filament/backend/src/opengl/PlatformCocoaGL.mm:
```diff
@@ -122,14 +129,27 @@
+            needsClear = true;
+        }
+        pImpl->updateOpenGLContext(nsView, true, needsClear);
     } else if (!CGRectEqualToRect(currentBounds, pImpl->mPreviousBounds)) {
```
Does the `mPreviousBounds` field need to exist?  Seems like this is duplicated data, since the conditional here is equivalent to: `!CGRectEqualToRect(currentBounds, pImpl->mPreviousSwapChainBounds[nsView])`.

prewettg replied:
I believe this was for the case when the context did not need to switch views, but the view had changed size. However, it is no longer needed after the refactor.

prideout commented:
Thanks!  It might be useful to add a multi-window app to `samples/` to demonstrate the issue that is being fixed here.

pixelflinger commented on filament/backend/src/opengl/PlatformCocoaGL.mm:
```diff
@@ -28,6 +28,7 @@
 #include <OpenGL/OpenGL.h>
 #include <Cocoa/Cocoa.h>
 
+#include <map>
```
use unordered_map instead

prewettg replied:
No longer applicable because of the refactor below.

pixelflinger commented on filament/backend/src/opengl/PlatformCocoaGL.mm:
```diff
@@ -38,8 +39,9 @@
     std::vector<NSView*> mHeadlessSwapChains;
-    NSRect mPreviousBounds = {};
-    void updateOpenGLContext(NSView *nsView, bool resetView);
+    std::map<NSView*, NSRect> mPreviousSwapChainBounds;
```
1. use unordered_map instead
2. we in fact don't need a map at all. But the implementation of SwapChain needs to change. Currently, we're simply casting the nsView to a Platform::SwapChain -- instead, we need to allocate a small CocoaGLSwapChain that stores the nsView and the bounds, and return that (and of course destroy that small structure when the swapchain is destroyed)

prideout replied:
Agreed with @pixelflinger.  @prewettg can you make these changes?

prewettg replied:
@prideout Sure, will do. Do you have a preference on which of the options to go with?  Option 2, changing SwapChain, seems like it would simplify things a lot, so I'm inclined to go with that one, but can do whichever works best for the project.

prideout replied:
Thanks!  We definitely want option 2.  :)

prewettg replied:
Implemented option 2. I found some additional problems with our (Open3D) case, when resizing. The resizing problems don't seem to happen with the SDL sample I added, it may be related to our use of GLFW instead of SDL.

pixelflinger reviewed (changes requested):
The use of a map<> shouldn't be needed. The more correct solution is slightly more complex (but will be tons more efficient and save binary size).

prideout commented:
Thank you, we will review this soon.

prewettg commented:
I added a sample as requested. I didn't use FilamentApp as this test is actually the guts, and it seemed simplest (from a reading-code perspective) to take just the parts from FilamentApp that are required to implement multiple windows. I implemented using SDL, since that's how FilamentApp does it. However, GLFW seems to have different behavior (at least within the Open3D framework). I'm not sure if there should be a GLFW version, or if that is out of scope, so I went with the latter.

pixelflinger commented on filament/backend/src/opengl/PlatformCocoaGL.mm:
```diff
@@ -34,12 +34,17 @@
 
 using namespace backend;
 
+struct CocoaGLSwapChain {
```
```suggestion
struct CocoaGLSwapChain : public Platform::SwapChain {
```

pixelflinger commented on filament/backend/src/opengl/PlatformCocoaGL.mm:
```diff
@@ -86,50 +91,90 @@
     }
 
-    return (SwapChain*) nativewindow;
+    return (Platform::SwapChain*)swapChain;
```
```suggestion
    return swapChain;
```

pixelflinger commented on filament/backend/src/opengl/PlatformCocoaGL.mm:
```diff
@@ -86,50 +91,90 @@
+    swapChain->previousBounds = NSZeroRect;
+    swapChain->previousWindowFrame = NSZeroRect;
+
+    return (Platform::SwapChain*)swapChain;
```
```suggestion
    return swapChain;
```

pixelflinger commented on filament/backend/src/opengl/PlatformCocoaGL.mm:
```diff
@@ -86,50 +91,90 @@
 }
 
 void PlatformCocoaGL::destroySwapChain(Platform::SwapChain* swapChain) noexcept {
+    CocoaGLSwapChain* cocoaSwapChain = (CocoaGLSwapChain*)swapChain;
```
```suggestion
    CocoaGLSwapChain* cocoaSwapChain = static_cast<CocoaGLSwapChain*>(swapChain);
```

pixelflinger reviewed (changes requested):
great, thanks! I suggested some changes to get better type safety.

prewettg commented:
@pixelflinger Sorry, I never actually checked Platform::SwapChain and was still in the mental mode of casting from the original code. Fixed to inherit from Platform::SwapChain.

pixelflinger commented:
This doesn't seem to build on Windows:
```
sdl2main.lib(SDL_windows_main.obj) : error LNK2019: unresolved external symbol SDL_main referenced in function main_getcmdline [D:\a\filament\filament\out\cmake-mt\samples\multiple_windows.vcxproj]
```
